### PR TITLE
fix: update utf-8 encoding and decoding implementations

### PR DESCRIPTION
**Pull request by backrunner**
To fix: https://github.com/FRSOURCE/base64/issues/88
The new implementations can handle the 4 bytes unicode emoji characters.

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -87,28 +79,10 @@ export const decode = (input: string) => {
...
-      utftext +=
-        String.fromCharCode((c >> 12) | 224) +
-        String.fromCharCode(((c >> 6) & 63) | 128) +
-        String.fromCharCode((c & 63) | 128);
-    }
-  }
-
-  return utftext;
+function _utf8_encode(str: string): string {
+  return encodeURIComponent(str).replace(/%([0-9A-F]{2})/g, (_, p1) =>
```
Comment (FRSgit) [inline index.ts:83]: Hm, I'm not sure about using the `encodeURIComponent` method - it had some gimmicks which blocked me from using it in the first place.
Pity I didn't write tests for that, give me few days to come up with reproduction of the issue I'm talking about - if I don't find it, we can go with your change

File: index.ts [hunk 2]
```diff
@@ -117,28 +91,17 @@ function _utf8_encode(string: string) {
...
+    return decodeURIComponent(
+      str
+        .split('')
+        .map((c) => '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2))
+        .join('')
+    );
+  } catch (error) {
+    // eslint-disable-next-line no-console
+    console.error('Failed to decode utf-8 base64 string: ', error);
+    return str;
```
Comment (FRSgit) [inline index.ts:105]: IMO this library is so tiny (it's a single util) that it's better to just throw an error when it happens. At least user can catch the error to react on it.
Returning `str` might lead to unhandled errors - when user expects that the string was decoded whilst in fact, it was not. With your implementation, there is simply no way to tell from the outside code the string was not decoded.

I'd recommend removing try/catch block whatsoever

**Comment by FRSgit** [comment]
Hi @backrunner!

Thanks for the contribution - it's great to see that you've reacted on the issue you've found! ❤️ 

As I wrote in another comment - I think there was a specific corner case that I handled in my code, I need to find it once again. Also, I plan to add vitest test suite on the main branch, because your case should be handled in the tests to make sure we won't make the same mistake again in the future. I'll try to add it tomorrow.

And last, but not least - please run `pnpm fix` to fix linting issues ✅

File: index.ts (hunk 2 above)
Comment (backrunner) [reply index.ts]: I accept that and just removed the try/catch block, and I noticed that my ESLint plugin is not working, sorry for the lint issue and they has been fixed too :D.

File: index.ts (hunk 1 above)
Comment (backrunner) [reply index.ts:83]: It will finally get the same result as TextEncoder, and if acceptable, use TextEncoder and TextDecoder can lead to a smaller bundle size.
For better compatibility I chose the `encodeURIComponent`, it will also use utf-8 inside to encode string, the 4 bytes unicode character will be converted into something like `%aa%bb%cc%dd`, when we converted it into 4 separated character, it will be the same thing as what TextEncoder.encode does at bytes level.

Comment (FRSgit) [reply index.ts:83]: Okay, so:

This one case that I was talking about - it was about ability to base64 encode/decode strings that are outside of standard Latin1 character range (e.g. special characters with diacritics like ąęś or just non-latin characters like 码). But your implementation handles those correctly as you're not using `atob` or `btoa` functions that are known for not handling special characters well 🎉 

I've added unit tests on the main branch, so please rebase your branch to get them. I've already prepared for you a test case [covering the issue you've reported](https://github.com/FRSOURCE/base64/blob/main/index.spec.ts#L27) (it's commented-out for now, so please just uncomment it).

Once the tests will run successfully on your branch I will happily merge this PR ❤️

Comment (backrunner) [reply index.ts:83]: > Okay, so:
> 
> This one case that I was talking about - it was about ability to base64 encode/decode strings that are outside of standard Latin1 character range (e.g. special characters with diacritics like ąęś or just non-latin characters like 码). But your implementation handles those correctly as you're not using `atob` or `btoa` functions that are known for not handling special characters well 🎉
> 
> I've added unit tests on the main branch, so please rebase your branch to get them. I've already prepared for you a test case [covering the issue you've reported](https://github.com/FRSOURCE/base64/blob/main/index.spec.ts#L27) (it's commented-out for now, so please just uncomment it).
> 
> Once the tests will run successfully on your branch I will happily merge this PR ❤️

I just rebased my branch and enable all the tests, they were all passed, you can check it later when you're available ❤

**Comment by FRSgit** [comment]
Nice, thanks for such great collaboration! I'm merging the PR and hope we'll e-meet each other again! 👋
